### ISSUE TITLE
About removal of categorical features support

### ISSUE THREAD

**Issue by ekerazha** [Microsoft/LightGBM#319]
I think the removal of categorical features support has been a really bad choice.
One-hot encoding is extremely expensive (computation time, memory usage), even in sparse format, if we have thousands of categories (I'm not talking about lightgbm itself, I'm talking about the data pre-processing). Categorical features support in lightgbm allowed to manage columns with thousands of categories in a nice way, with a small pre-processing cost and lean memory usage. Now this isn't possible anymore.

**Comment by guolinke** [comment]
The previous support of categorical features will slow down the training speed a little.
If this is really needed, I will figure out a new way to enable it again.

Thanks

**Comment by ekerazha** [comment]
Ok, thanks.
For my part, I'll keep using the latest commit with categorical features support.

**Comment by Laurae2** [comment]
@guolinke What do you think about the following?:

* Passing `categorical_features` (or providing `-1`) disables Exclusive Feature Bundling
* Not passing `categorical_features` enables Exclusive Feature Bundling

Another solution is user should stick to v1 for `categorical_features` support, but it is not the best solution for the end-users.

**Comment by guolinke** [comment]
@Laurae2 
The support of categorical feature and exclusive features bundling does not conflict with each other.
They can work at the same time.

**Comment by Laurae2** [comment]
@guolinke I meant sometimes there are slowness issues with EFB (hence the reason of giving the user the ability to disable it, if provided a specific input). Here is an example of a 70x30k matrix, which takes longer to perform EFB than training (some seconds):

![image](https://cloud.githubusercontent.com/assets/9083669/23339152/2cd444aa-fc1c-11e6-9a9a-b24061627ed3.png)

As that matrix is reconstructed (differently) after training, it means it spends more time grouping data together than training (but it does work properly - it also reveals data properties we might not have found without it).

**Comment by guolinke** [comment]
@Laurae2 
OK, I see. you can use ```enable_bundle=false``` to disable it .

**Comment by ekerazha** [comment]
@guolinke I'm trying v2 with categorical features support, I'm using the same code I used with v1, but something goes wrong with v2:

```
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=33
[43]	valid_0's l2: 7200.34
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=21
[44]	valid_0's l2: 7137.16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=21
[45]	valid_0's l2: 7075.31
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=23
[46]	valid_0's l2: 7014.71
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=28
[47]	valid_0's l2: 6955.29
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=19
[48]	valid_0's l2: 2.17785e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=31
[49]	valid_0's l2: 2.17752e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=34
[50]	valid_0's l2: 2.17358e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=325
[51]	valid_0's l2: 2.17005e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=29
[52]	valid_0's l2: 2.16975e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=25
[53]	valid_0's l2: 2.16934e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=20
[54]	valid_0's l2: 2.1684e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=237
[55]	valid_0's l2: 2.16645e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=26
[56]	valid_0's l2: 2.16554e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=27
[57]	valid_0's l2: 2.16534e+16
[LightGBM] [Info] Trained a tree with leaves=1023 and max_depth=23
```

Look at `[48]`, at some point the validation loss just explodes.

**Comment by guolinke** [comment]
can you provide the code and a sampling data so that i can reproduce this?

**Comment by ekerazha** [comment]
Can I send an e-mail to the address you have on your github profile (i at yumumu)?

**Comment by guolinke** [comment]
@ekerazha Sure

**Comment by guolinke** [comment]
@ekerazha can you try the latest commit?

**Comment by ekerazha** [comment]
It looks like the bug has been fixed, although with the same hyperparameters it performs a little bit worse than v1 on my dataset (RMSE is a little bit higher, but it's a tiny difference).

**Comment by AbdealiJK** [comment]
The categorical feature support has been extremely useful for my work! Thanks a lot for adding it back :)
